### PR TITLE
refactor!: height-gated hard fork upgrade to prevent staking power changes

### PR DESCRIPTION
**Pull request by alexanderbez**
## Description

This patch prevents messages that would modify validator staking power on Terra. Note, staking power can still change due to delegation of rewards, but only on existing validators. 

There exists a const, `StakingPowerUpgradeHeight`, that defines the height (yet to be agreed upon), after which such messages would not longer be accepted by Terra's state machine.

Pending acceptance, this PR should be merged with a new tag being released, `v0.44.5-terra.3`, along with a corresponding release of terra-core, which at least 2/3 of validators must upgrade to in a coordinated manner.

TODO:

- [x] Pick a height for `StakingPowerUpgradeHeight`

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/staking/keeper/msg_server.go [hunk 1]
```diff
@@ -14,6 +14,10 @@ import (
 	"github.com/cosmos/cosmos-sdk/x/staking/types"
 )
 
+// StakingPowerUpgradeHeight defines the block height after which messages that
+// would impact staking power are no longer supported.
+const StakingPowerUpgradeHeight = 0
```
Comment (ValarDragon) [inline x/staking/keeper/msg_server.go:19]:
```suggestion
const StakingPowerUpgradeHeight = 7603700
```

**Comment by alexanderbez** [comment]
> Overall looks good.
> 
> The only other attack I could see is if someone surreptitiously was making delegations to various validators and then began to collect them to a single validator via redelegations.

That single malicious validator would already have to exist on chain though, but yeah.

**Comment by joeabbey** [comment]
> > Overall looks good.
> > The only other attack I could see is if someone surreptitiously was making delegations to various validators and then began to collect them to a single validator via redelegations.
> 
> That single malicious validator would already have to exist on chain though, but yeah.

🤞 I hope that's highly unlikely ... but we are sort of in hypothetical wargame-mode 🤷

**Comment by alexanderbez** [comment]
This PR has nothing to do with any proposals. In fact, proposals were only brought up after this made made and released. This PR solely addresses an attack vector after the depeg.

**Comment by Raider7019** [comment]
This patch was in reality the actual attack vector! Ensuring that 370 million LUNC can now override the wishes of 6.5 trillion community LUNC holders, leaving the community disenfranchised to the benefit of TFL. 

The "governance attack" you are talking about is really TFL's desire to remain in control and not be outvoted. Instead of this patch TFL could have bought up LUNC in the market to ensure they retained a legitimate majority.

So far @immibis was spot on with his comments. I hope you devs push for the urgent development and installation by validators of a new patch that reverts the disabling of delegation and validator node registration to allow the Terra Classic community to move forward instead of this limbo state we find ourselves in.

Rant over ;)

**Comment by aeuser999** [comment]
I concur with Raider7019.  [Documentation](http://classic-docs.terra.money/docs/learn/protocol.html#governance) specifically states that protocol changes are made by governance proposals that have passed, and that governance is decentralized.  This change was not made in accordance with those rules.  

I also ask that a new patch be immediately considered in order to revert the disabling of delegation and validator node registration for Terra classic (and stay in keeping with the intent behind "Once... what is being termed as Terra Classic launches, this will be reverted obviously", at least for delegation and validator node registration).

**Pull request merged**